Signal/src/Profiles/OWSProfileManager.m

```diff
@@ -987,76 +923,80 @@ - (BOOL)isNullableStringEqual:(NSString *_Nullable)left toString:(NSString *_Nul
 {
-    OWSAssert(profileKey.length == kProfileKeyLength);
+    OWSAssert(profileKey.keyData.length == kAES128_KeyByteLength);
 
-    if (!data) {
+    NSData *_Nullable decryptedData = [self decryptProfileData:encryptedData profileKey:profileKey];
+    if (decryptedData.length < 1) {
         return nil;
     }
 
-    // TODO: Encrypt.  For now, return the input.
-    return data;
-}
 
-+ (NSData *_Nullable)encryptProfileString:(NSString *_Nullable)value profileKey:(NSData *)profileKey
-{
-    OWSAssert(profileKey.length == kProfileKeyLength);
+    // Unpad profile name.
+    NSUInteger unpaddedLength = 0;
+    const char *bytes = decryptedData.bytes;
 
-    if (value) {
-        NSData *_Nullable data = [value dataUsingEncoding:NSUTF8StringEncoding];
-        if (data) {
-            NSData *_Nullable encryptedData = [self encryptProfileData:data profileKey:profileKey];
-            return encryptedData;
+    // Work through the bytes until we encounter our first
+    // padding byte (our padding scheme is NULL bytes)
+    for (NSUInteger i = 0; i < decryptedData.length; i++) {
+        if (bytes[i] == 0x00) {
+            break;
         }
+        unpaddedLength = i + 1;
     }
 
-    return nil;
-}
+    NSData *unpaddedData = [decryptedData subdataWithRange:NSMakeRange(0, unpaddedLength)];
 
-- (NSData *_Nullable)decryptProfileData:(NSData *_Nullable)encryptedData profileKey:(NSData *)profileKey
-{
-    return [OWSProfileManager decryptProfileData:encryptedData profileKey:profileKey];
+    return [[NSString alloc] initWithData:unpaddedData encoding:NSUTF8StringEncoding];
 }
 
-- (NSString *_Nullable)decryptProfileString:(NSData *_Nullable)encryptedData profileKey:(NSData *)profileKey
+- (nullable NSData *)encryptProfileData:(nullable NSData *)data
 {
-    return [OWSProfileManager decryptProfileString:encryptedData profileKey:profileKey];
+    return [self encryptProfileData:data profileKey:self.localProfileKey];
 }
 
-- (NSData *_Nullable)encryptProfileData:(NSData *_Nullable)data
+- (nullable NSData *)encryptProfileNameWithUnpaddedName:(NSString *)name
 {
-    return [OWSProfileManager encryptProfileData:data profileKey:self.localProfileKey];
-}
+    if (name.length == 0) {
+        return nil;
+    }
+    
+    NSData *nameData = [name dataUsingEncoding:NSUTF8StringEncoding];
+    if (nameData.length > kOWSProfileManager_NameDataLength) {
+        OWSFail(@"%@ name data is too long with length:%lu", self.tag, (unsigned long)nameData.length);
+        return nil;
+    }
 
-- (NSData *_Nullable)encryptProfileString:(NSString *_Nullable)value
-{
-    return [OWSProfileManager encryptProfileString:value profileKey:self.localProfileKey];
+    NSUInteger paddingByteCount = kOWSProfileManager_NameDataLength - nameData.length;
+
+    NSMutableData *paddedNameData = [nameData mutableCopy];
+    [paddedNameData increaseLengthBy:paddingByteCount];
```

Comment:
I would add a comment that `increaseLengthBy:` sets all of the added bytes to zero.